coderd/provisionerdserver/acquirer.go

```diff
@@ -0,0 +1,481 @@
+			if internalError != nil {
+				// internalError takes precedence
+				return database.ProvisionerJob{}, internalError
+			}
+			return database.ProvisionerJob{}, err
+		case <-clearance:
+			logger.Debug(ctx, "got clearance to call database")
+			job, err := a.store.AcquireProvisionerJob(principal, database.AcquireProvisionerJobParams{
+				StartedAt: sql.NullTime{
+					Time:  dbtime.Now(),
+					Valid: true,
+				},
+				WorkerID: uuid.NullUUID{
+					UUID:  worker,
+					Valid: true,
+				},
+				Types: pt,
+				Tags:  dbTags,
+			})
+			if xerrors.Is(err, sql.ErrNoRows) {
+				logger.Debug(ctx, "no job available")
+				continue
+			}
+			// we are not going to retry, so signal we are done
+			internalError := a.done(dk, clearance)
+			if internalError != nil {
+				// internal error takes precedence
+				return database.ProvisionerJob{}, internalError
+			}
+			if err != nil {
+				logger.Warn(ctx, "error attempting to acquire job", slog.Error(err))
+				return database.ProvisionerJob{}, xerrors.Errorf("failed to acquire job: %w", err)
+			}
+			logger.Debug(ctx, "successfully acquired job")
+			return job, nil
+		}
+	}
+}
+
+// want signals that an acquiree wants clearance to query for a job with the given dKey.
+func (a *Acquirer) want(pt []database.ProvisionerType, tags Tags, clearance chan<- struct{}) {
+	dk := domainKey(pt, tags)
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	cleared := false
+	d, ok := a.q[dk]
+	if !ok {
+		ctx, cancel := context.WithCancel(a.ctx)
+		d = domain{
+			ctx:       ctx,
+			cancel:    cancel,
+			a:         a,
+			key:       dk,
+			pt:        pt,
+			tags:      tags,
+			acquirees: make(map[chan<- struct{}]*acquiree),
+		}
+		a.q[dk] = d
+		go d.poll(a.backupPollDuration)
+		// this is a new request for this dKey, so is cleared.
+		cleared = true
+	}
+	w, ok := d.acquirees[clearance]
+	if !ok {
+		w = &acquiree{clearance: clearance}
+		d.acquirees[clearance] = w
+	}
+	// pending means that we got a job posting for this dKey while we were
+	// querying, so we should clear this acquiree to retry another time.
+	if w.pending {
+		cleared = true
+		w.pending = false
+	}
+	w.inProgress = cleared
+	if cleared {
+		// this won't block because clearance is buffered.
+		clearance <- struct{}{}
+	}
+}
+
+// cancel signals that an acquiree no longer wants clearance to query.  Any error returned is a serious internal error
+// indicating that integrity of the internal state is corrupted by a code bug.
+func (a *Acquirer) cancel(dk dKey, clearance chan<- struct{}) error {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	d, ok := a.q[dk]
+	if !ok {
+		// this is a code error, as something removed the dKey early, or cancel
+		// was called twice.
+		err := xerrors.New("canceled non-existent job acquisition")
+		a.logger.Critical(a.ctx, "internal error", slog.Error(err))
+		return err
+	}
+	w, ok := d.acquirees[clearance]
+	if !ok {
+		// this is a code error, as something removed the dKey early, or cancel
+		// was called twice.
```

Comment:
copy-pasta?

Reply:
the state is a 2-level heirarchy of domain then acquiree.  We clean up the domain when there are no acquirees, so calling `cancel` twice could cause this error or the one on L207 depending on whether there are other acquirees in the domain.

Reply:
Ah, understood.